File: packages/dtj-typescript/src/discovery.ts
/**
 * Agent discovery and process management.
 * Handles finding dtj-agent binary and managing its lifecycle.
 */

import { spawn, ChildProcess } from "node:child_process";
import { existsSync, mkdirSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { mkdtempSync } from "node:fs";
import { DTJAgentNotFoundError, DTJConnectionError } from "./errors.js";

export interface AgentDiscoveryOptions {
  agentPath?: string;
  socketPath?: string;
  dataDir?: string;
}

export class AgentDiscovery {
  private agentPath?: string;
  private socketPath?: string;
  private dataDir: string;
  private process: ChildProcess | null = null;
  private tempSocketDir: string | null = null;
  private warningEmitted = false;

  constructor(options: AgentDiscoveryOptions = {}) {
    this.agentPath = options.agentPath;
    this.socketPath = options.socketPath;
    this.dataDir = options.dataDir || join(process.cwd(), "traces");
  }

  /**
   * Find dtj-agent binary using discovery order:
   * 1. Explicit agentPath
   * 2. DTJ_AGENT_PATH environment variable
   * 3. PATH lookup for 'dtj-agent'
   */
  findAgent(): string | null {
    // 1. Explicit agent_path
    if (this.agentPath) {
      if (existsSync(this.agentPath)) {
        return this.agentPath;
      }
      // Don't throw - return null to trigger no-op session with warning
      return null;
    }

    // 2. DTJ_AGENT_PATH environment variable
    const envPath = process.env["DTJ_AGENT_PATH"];
    if (envPath && existsSync(envPath)) {
      return envPath;
    }

    // 3. PATH lookup
    const pathDirs = process.env["PATH"]?.split(":") || [];
    for (const dir of pathDirs) {
      const candidate = join(dir, "dtj-agent");
      if (existsSync(candidate)) {
        return candidate;
      }
    }

    return null;
  }

  /**
   * Start dtj-agent and return socket path.
   * If socketPath provided, don't start agent (connect to existing).
   */
  startAgent(): string {
    // If socket_path provided, don't start agent (connect to existing)
    if (this.socketPath) {
      return this.socketPath;
    }

    const agentBinary = this.findAgent();
    if (!agentBinary) {
      this.emitWarning();
      throw new DTJAgentNotFoundError(
        "dtj-agent not found. Install dtj-agent or set DTJ_AGENT_PATH. Tracing disabled."
      );
    }

    // Create temporary directory for socket
    this.tempSocketDir = mkdtempSync(join(tmpdir(), "dtj-agent-"));
    const socketPath = join(this.tempSocketDir, "agent.sock");

    // Ensure data_dir exists
    mkdirSync(this.dataDir, { recursive: true });

    // Start agent process
    try {
      this.process = spawn(agentBinary, ["--socket", socketPath, "--data-dir", this.dataDir], {
        stdio: "ignore",
        detached: false,
      });

      // Handle process errors
      this.process.on("error", (err) => {
        console.error(`dtj-agent process error: ${err.message}`);
      });

      this.process.on("exit", (code, signal) => {
        if (code !== 0 && code !== null) {
          console.error(`dtj-agent exited with code ${code}, signal ${signal}`);
        }
      });
    } catch (err) {
      this.cleanup();
      throw new DTJConnectionError(`Failed to start dtj-agent: ${err}`);
    }

    return socketPath;
  }

  /**
   * Stop the agent process and cleanup.
   */
  stopAgent(): void {
    if (this.process) {
      try {
        this.process.kill("SIGTERM");
        // Wait for process to exit (with timeout)
        const exitPromise = new Promise<void>((resolve) => {
          const timeout = setTimeout(() => {
            if (this.process) {
              this.process.kill("SIGKILL");
            }
            resolve();
          }, 5000);
          this.process?.on("exit", () => {
            clearTimeout(timeout);
            resolve();
          });
        });
        // Don't await - fire and forget for cleanup
        exitPromise.catch(() => {});
      } catch {
        // Ignore errors during cleanup
      } finally {
        this.process = null;
      }
    }

    this.cleanup();
  }

  /**
   * Cleanup temporary resources.
   */
  private cleanup(): void {
    if (this.tempSocketDir) {
      try {
        // Remove temp directory and contents
        rmSync(this.tempSocketDir, { recursive: true, force: true });
      } catch {
        // Ignore cleanup errors
      }
      this.tempSocketDir = null;
    }
  }

  /**
   * Emit single Node warning about disabled tracing.
   */
  private emitWarning(): void {
    if (!this.warningEmitted) {
      process.emitWarning(
        "dtj-sdk: tracing disabled (no dtj-agent found). " +
        "Install dtj-agent or set DTJ_AGENT_PATH to enable.",
        {
          type: "DTJWarning",
          code: "DTJ_AGENT_UNAVAILABLE",
        }
      );
      this.warningEmitted = true;
    }
  }

  /**
   * Get the socket path (for connecting to existing agent).
   */
  getSocketPath(): string | undefined {
    return this.socketPath;
  }
}

File: packages/dtj-python/src/dtj_sdk/discovery.py
"""Agent discovery and process management."""

import os
import shutil
import subprocess
import tempfile
import warnings
from pathlib import Path
from typing import Optional

from .exceptions import DTJAgentNotFoundError, DTJConnectionError


class AgentDiscovery:
    """Handles discovery and lifecycle of dtj-agent binary."""
    
    def __init__(
        self,
        agent_path: Optional[str] = None,
        socket_path: Optional[str] = None,
        data_dir: Optional[str] = None,
        config_path: Optional[str] = None,
        use_config: bool = False,
    ):
        self.agent_path = agent_path
        self.socket_path = socket_path
        self.data_dir = data_dir
        self.config_path = config_path
        self.use_config = use_config
        self._process: Optional[subprocess.Popen] = None
        self._temp_socket_dir: Optional[tempfile.TemporaryDirectory] = None
        self._warning_emitted = False
    
    def find_agent(self) -> Optional[str]:
        """Find dtj-agent binary using discovery order."""
        # 1. Explicit agent_path
        if self.agent_path:
            path = Path(self.agent_path)
            if path.is_file() and os.access(path, os.X_OK):
                return str(path)
            raise DTJAgentNotFoundError(f"Specified agent_path not executable: {self.agent_path}")
        
        # 2. DTJ_AGENT_PATH environment variable
        env_path = os.environ.get("DTJ_AGENT_PATH")
        if env_path:
            path = Path(env_path)
            if path.is_file() and os.access(path, os.X_OK):
                return str(path)
        
        # 3. PATH lookup
        which_path = shutil.which("dtj-agent")
        if which_path:
            path = Path(which_path)
            if path.is_file() and os.access(path, os.X_OK):
                return str(path)
        
        # 4. macOS Homebrew fallback
        for candidate in ("/opt/homebrew/bin/dtj-agent", "/usr/local/bin/dtj-agent"):
            path = Path(candidate)
            if path.is_file() and os.access(path, os.X_OK):
                return str(path)
        
        # 5. Cargo dev install fallback
        cargo_path = Path.home() / ".cargo" / "bin" / "dtj-agent"
        if cargo_path.is_file() and os.access(cargo_path, os.X_OK):
            return str(cargo_path)
        
        return None
    
    def start_agent(self) -> str:
        """Start dtj-agent and return socket path."""
        # If socket_path provided, don't start agent (connect to existing)
        if self.socket_path:
            return self.socket_path
        
        agent_binary = self.find_agent()
        if not agent_binary:
            self._emit_warning()
            raise DTJAgentNotFoundError(
                "dtj-agent not found. Install dtj-agent or set DTJ_AGENT_PATH. "
                "Tracing disabled."
            )
        
        # Ensure data_dir exists - default to ./traces if not set
        data_dir = Path(self.data_dir) if self.data_dir else Path.cwd() / "traces"
        
        if self.use_config and self.config_path:
            # Use --config flag instead of --data-dir
            config_file = Path(self.config_path)
            if not config_file.is_file():
                # Try to read it anyway - start_agent will handle the error
                pass
            
            # Create socket dir for agent (even when using config)
            self._temp_socket_dir = tempfile.TemporaryDirectory(prefix="dtj-agent-")
            socket_dir = Path(self._temp_socket_dir.name)
            socket_path = socket_dir / "agent.sock"
            
            # Start agent with --config flag instead of --data-dir
            try:
                self._process = subprocess.Popen(
                    [agent_binary, "--socket", str(socket_path), "--config", str(config_file)],
                    stdin=subprocess.DEVNULL,
                    stdout=subprocess.DEVNULL,
                    stderr=subprocess.DEVNULL,
                )
            except Exception as e:
                self._cleanup()
                raise DTJConnectionError(f"Failed to start dtj-agent with --config: {e}")
            
            return str(socket_path)
        
        # Create temporary directory for socket (default behavior)
        self._temp_socket_dir = tempfile.TemporaryDirectory(prefix="dtj-agent-")
        socket_dir = Path(self._temp_socket_dir.name)
        socket_path = socket_dir / "agent.sock"
        
        # Ensure data_dir exists (use --data-dir flag)
        data_dir.mkdir(parents=True, exist_ok=True)
        
        # Start agent process with --data-dir flag  
        try:
            self._process = subprocess.Popen(
                [agent_binary, "--socket", str(socket_path), "--data-dir", str(data_dir)],
                stdin=subprocess.DEVNULL,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        except Exception as e:
            self._cleanup()
            raise DTJConnectionError(f"Failed to start dtj-agent: {e}")
        
        return str(socket_path)
    
    def stop_agent(self) -> None:
        """Stop the agent process and cleanup."""
        if self._process:
            try:
                self._process.terminate()
                self._process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._process.kill()
                self._process.wait()
            except Exception:
                pass
            finally:
                self._process = None
        
        self._cleanup()
    
    def _cleanup(self) -> None:
        """Cleanup temporary resources."""
        if self._temp_socket_dir:
            try:
                self._temp_socket_dir.cleanup()
            except Exception:
                pass
            self._temp_socket_dir = None
    
    def _emit_warning(self) -> None:
        """Emit single RuntimeWarning about disabled tracing."""
        if not self._warning_emitted:
            warnings.warn(
                "dtj-agent not found; tracing disabled (no-op mode). "
                "Install dtj-agent or set DTJ_AGENT_PATH to enable.",
                RuntimeWarning,
                stacklevel=3,
            )
            self._warning_emitted = True
    
    def __enter__(self) -> "AgentDiscovery":
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.stop_agent()